File: cpp/leetcode/356.cpp
bool isReflected(vector<pair<int, int>>& points) {
    int len = points.size();
    if (len==0 || len==1) return true;
    
    unordered_map<int, set<int>> mp;
    int max = points[0].first, min = points[0].first;
    // build the map, find min and max
    for (int i=0; i<len; ++i) {
        if (points[i].first < min) min = points[i].first;
        if (points[i].first > max) max = points[i].first;
        mp[points[i].second].insert(points[i].first);
    }
    double line = (min + max) / 2.0;
    
    // mirror the sorted x value in the set using two pointer
    for (auto it = mp.begin(); it!=mp.end(); ++it) {
        set<int>& s = it->second;
        for (auto start=s.begin(), end = s.end(); start!=end; ++start) {
            if ((*start + *(--end)) / 2.0 != line)
                return false;
            if (start==end) break;
        }
    }
    return true;
}

File: python/leetcode/356.py
class Solution(object):
    def isReflected(self, points):
        """
        :type points: List[List[int]]
        :rtype: bool
        group points by y positions, then check x in pairs
        """
        groups = {}
        for x, y in sorted(points):
            if y not in groups:
                groups[y] = [x]
            else:
                groups[y] += x,

        possible = None
        for ls in groups.values():
            l, r = 0, len(ls) - 1
            while l <= r:
                tmp = ls[l] + ls[r]
                if tmp % 2:
                    return False
                tmp /= 2
                if possible and possible != tmp:
                    return False
                possible = tmp
                l += 1
                r -= 1
        return True


    def isReflected(self, points):
        points.sort()
        return points == sorted([points[0][0] + points[-1][0] - x, y]
                                for x, y in points)


    def isReflected(self, points):
        if not points: return True
        X = min(points)[0] + max(points)[0]
        return {(x, y) for x, y in points} == {(X - x, y) for x, y in points}